teps/0115-tekton-catalog-git-based-versioning.md

```diff
@@ -0,0 +1,398 @@
+that seem minor cause surprising breaking changes for users, such as described in [tektoncd/catalog#784][784].
+- Resources that are frequently updated with new versions end up with a lot of duplications because any change requires
+copying over everything into a new directory. For example, [git-clone][git-clone] `Task` has seven versions with a lot
+of duplication across versions.
+
+Previously, there was tight coupling between the organization and how users referenced resources which made it hard to 
+change the organization of the Catalog without breaking users. In [TEP-0110][tep-0110], we decoupled the Catalog 
+organization from resource reference. Now, we can create a new organization and versioning contract for Catalogs to
+address the challenges with the directory-based organization and versioning.
+
+The aim of this TEP is to create a new organization and versioning for Tekton Catalogs, building on the prior work and 
+experiences from [TEP-0003][tep-0003]. In [TEP-0079: Tekton Catalog Support Tiers][tep-0079], we plan to add a Tekton 
+Official Catalog which will use this new organization and versioning contract - therefore, this proposal is a blocker 
+for [TEP-0079][tep-0079].
+
+## Proposal
+
+We propose supporting a flat organization and git-based versioning of Tekton Catalogs.
+
+### Flat Organization
+
+In this organization contract, we'll use a flat structure instead of the directory structure. The example shown above
+with the two `Tasks` (argocd and golang-build) and one `Pipeline` (go-release) would be organized as such:
+
+```shell
+./task/
+  /argocd
+    /README.md
+    /argocd.yaml
+    /samples
+    /tests
+    /OWNERS
+  /golang-build
+    /README.md
+    /golang-build.yaml
+    /samples
+    /tests
+    /OWNERS
+./pipeline/
+  /go-release
+    /README.md
+    /go-release.yaml
+    /samples
+    /tests
+    /OWNERS
+```
+
+Note that there are no changes to the contents of the resources themselves, only the organization of Catalogs.
+
+### Git-Based Versioning
+
+#### Versioning via Tags
+
+Versioning of resources will be done through creating tags in the Catalogs. The versions in the Tekton Hub should
+continue to use the simple version of [semver][semver]: `{major}.{minor}`. We propose that the tags use the semantic
+versioning model - `v<major-version>.<minor-version>`. Minor versions are made to signal backwards-compatible changes,
+while major versions are made to signal backwards-incompatible changes. As such, the tags in the catalogs will map to
+versions in the Hub e.g. tag "v0.7" in a Catalog maps to version "0.7" in the Hub.
+
+#### One Version per Catalog
+
+Note that version tags apply to the whole repository. Therefore, we propose that Catalogs should contain one or
+multiple resource that have the same versioning. Instead of the centralized Catalog from the community, we will
+encourage users to create Catalogs in their own GitHub organizations and share them through the Hub.
```

Comment:
nomenclature check: I was imagining the difference between hub and catalog being:

- hub: UI frontend for a catalog
- catalog: A collection of repositories supplying Tasks (alt - hub backend)
- repository: versioned source containing Tasks

So we'd still have a Tekton catalog, but it would be backed by user repositories as opposed to the current monorepo.

where as this definition it sounds like the hub is a collection of catalogs, with catalogs mapping to source repositories. 

wdyt?

Reply:
@wlynch these are the definitions from my perspective:

- Hub: UI frontend for one or multiple catalogs
- Catalog: repository of one or multiple Tekton resources that meets an organization contract (for Hub to interpret)

there's no special definition for "repository" in the context of this TEP

Reply:
@wlynch added these definitions and clarifications to the TEP

Reply:
I think this may cause some naming mismatch / confusion - the phrase I hear most often is w.r.t. to "the catalog" (e.g. "find a task in the catalog", "add a task to the catalog", etc.) which I suspect will be a very difficult habit to break for both users and maintainers.

That said, this is minor so I don't want to get hung up on it (definitely don't consider this blocking - we can always change it later), but worth considering before we start assigning names to types.